Integration test skeleton

pablochacin opened the pull request:
This is an initial version of the test code. It is not yet integrated into CI test.

**What this PR does / why we need it**:

Complements [PR-755](https://github.com/kubernetes-sigs/cluster-api/pull/755) implementing an skeleton for integration tests based on events.

Fixes #11 

TODO
- [x] Implement skeleton
- [x] Test locally
- [x] Integrate into ci scripts
- [x] Test in CI environment

**Special notes for your reviewer**:

Code submitted for review of the proposed test approach. 

cc @figo

ncdc commented on test/integration/cluster/cluster_test.go:
```diff
@@ -0,0 +1,105 @@
+					}
+				},
+			})
+			go informer.Run(stopper)
```
I would add `cache.WaitForCacheSync(stopper, informer.HasSynced)` to let the informer's cache fully synchronize before proceeding. There's a slim chance you could miss events, otherwise.

pablochacin replied:
Good point. implemented

ncdc commented on test/integration/cluster/cluster_test.go:
```diff
@@ -0,0 +1,105 @@
+			events := make(chan *corev1.Event, 0)
+			informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					e := obj.(*corev1.Event)
```
Do you want to check if this is an Event and fail if it's not?

pablochacin replied:
The intention of this skeleton is to show the general form of the tests: inform of events related to the cluster and check those are the expected events. The actual verification depends of each specific test. Also, it requires the events carry the required information, something that has still to be reviewed in the test provider.

ncdc replied:
If someone accidentally coded `informer = factory.Core().V1().Pods().Informer()` (pods instead of events), this type assertion would fail. I would say it's safer to validate the type of `obj`.

ncdc replied:
I still think it's worth validating the received type. Something like
```go
Expect(obj).Should(BeAssignableToTypeOf(&corev1.Event{}))
```
... if that's the correct gomega matcher syntax.

pablochacin replied:
@ncdc The informer is created for events: `informer = factory.Core().V1().Events().Informer()`.

I expect only Events to be received. Wouldn't checking type be redundant?

ncdc replied:
I agree with you that it is unlikely you would get an incorrect type, but it is technically possible due to programmer error (using the wrong informer). The informer framework does not have static typing guarantees, which is why I suggested adding the type check. It may sound redundant, but I'd rather have an `Expect` failure than a panic in the event of an accidental mismatch between the informer and the expected type. However, I'm not going to hold this PR up because of this 😄

pablochacin commented on test/integration/cluster/cluster_test.go:
```diff
@@ -0,0 +1,105 @@
+	var informer cache.SharedIndexInformer
+
+	BeforeEach(func() {
+
```
done

pablochacin commented on test/integration/cluster/cluster_test.go:
```diff
@@ -0,0 +1,105 @@
+var clusterInstance = &clusterv1alpha1.Cluster{
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      "foo",
+		Namespace: "clusterapi-test",
```
Definitely. It will prevent problems if a test starts before the namespace of the previous test is still terminating.

pablochacin commented on test/integration/cluster/cluster_test.go:
```diff
@@ -0,0 +1,105 @@
+
+var clusterInstance = &clusterv1alpha1.Cluster{
+	ObjectMeta: metav1.ObjectMeta{
+		Name:      "foo",
```
Given that each test will have its own namespace, I think it is not necessary. Neither does it harm.

figo commented on test/integration/cluster/cluster_test.go:
```diff
@@ -0,0 +1,105 @@
+			_, err := clusterapi.Create(cluster)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			Expect(<-events).NotTo(BeNil())
```
do you want to set a timeout for waiting?  will need to do non blocking channel polling instead.

pablochacin replied:
I Changed Expect by Eventually, and used a nonblocking channel receive.

figo commented:
/lgtm

figo commented:
/lgtm cancel
@pablochacin  you will need to run `./hack/update-bazel.sh ` to generate the BUILD.bazel in folder.
and add copyright header for the file.

pablochacin commented:
@figo I addressed your comments. Thanks for pointing out the copyright!
Regarding the `scripts/ci-test.sh` I cannot test it locally as I get an `OS linux not supported` message.

figo commented:
/ok-to-test

figo commented:
> @figo I addressed your comments. Thanks for pointing out the copyright! I also changed the package to `integration` (it was `deployment` not sure why)
> Regarding the `scripts/ci-test.sh` I cannot test it locally as I get an `OS linux not supported` message.-

do you know which line the error comes from?  running on linux is fine (ci is running with Linux container)

btw, the latest failure shows you need to push the updated vendor files
but please make sure `scripts/ci-test.sh`, `scripts/ci-build.sh`, `scripts/ci-make.sh` pass locally.

ncdc commented:
/milestone v1alpha1

figo commented on scripts/fetch_ext_bins.sh:
```diff
@@ -31,7 +31,7 @@ k8s_version=1.11.0
 goos="unknown"
 
-if [[ "$OSTYPE" == "linux-gnu" ]]; then
+if [[ "$OSTYPE" == "linux"* ]]; then
```
what the other linux we are trying to support beside `linux-gnu`?

pablochacin replied:
in my local machine, OSTYPE is just `linux`.

figo commented:
/lgtm

figo commented:
/assign @ncdc

ncdc commented on test/integration/cluster/cluster_test.go:
```diff
@@ -0,0 +1,139 @@
+	},
+}
+
+// Constructs a function for async event retrieval
```
godoc for functions requires that the first word of the comment match the name of the function. How about something like this?
```go
// checkEvents returns a function that attempts to retrieve an event from c. The returned function is compatible with Gomega's Eventually assertion.
```

pablochacin replied:
True. Thanks.

ncdc commented on test/integration/cluster/cluster_test.go:
```diff
@@ -0,0 +1,139 @@
+			_, err := clusterapi.Create(cluster)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			Eventually(checkEvents(events), TIMEOUT, RETRY).ShouldNot(BeNil())
```
We alternatively could consider either using this model (https://onsi.github.io/ginkgo/#asynchronous-tests) or making sure we set a reasonable `--timeout` value when running `ginkgo` (the default is 24hr) or `-timeout` for `go test` (the default is 10m).

figo replied:
+1 for ginkgo asynchronous-test.

pablochacin replied:
@ncdc @figo I already using asynchronous test for receiving the events in line 135 with `Eventually`.  What other asynchronous check are you suggesting?

pablochacin replied:
@ncdc @figo The test are already asynchronous. The informer notifies the test of events by means of channels and the test is done using `gomega`'s `Eventually` assertion.  Not sure what specific changes you expect.

ncdc replied:
It's not necessary, but I was trying to avoid a polling loop (which is what `Eventually` does). Something like this:

```go
It("should trigger an event", func(done Done) {
  // ...
  Expect(<-events).ShouldNot(BeNil())
  close(done)
}, TIMEOUT)
```

pablochacin replied:
@ncdc this was the patter of my initial code, but ti was suggested (IIRC by @figo) to avoid the blocking `Expect(<-events)` and I agree that doing it non-blocking channel receive was better. Therefore I moved to use `Eventually` and a non-blocking receive, which requieres some sort of polling.

ncdc replied:
If you use the code pattern I wrote above, it allows you to have a timeout with needing to poll.

Either approach should be fine. I was suggesting something that I would prefer (not polling + you don't have to write a `checkEvents` function), but if you prefer this way, 👍 and let's get this merged.

pablochacin replied:
I see, I agree this way is simpler.

ncdc commented:
I'm used to seeing commits that modify vendored dependencies separated from commits that modify this repository's source code. @detiber have you all been following that practice here?

detiber commented:
@ncdc If we wanted to enforce that here we'd also have to leverage separate PRs, since the default policy for the cluster-api* repos is to squash merge (I believe that is the default for at least kubernetes-sigs repos, if not all kubernetes managed repos)

ncdc commented:
Ok, just checking. Not a big deal.

figo commented on test/integration/cluster/cluster_test.go:
```diff
@@ -0,0 +1,139 @@
+	}
+}
+
+const TIMEOUT = 10 * time.Second
```
not sure about a perfect reasonable timeout value, 10s is not enough in my local testing. 1m should be good.
i still think it is good to have a timeout on this call even the test has a overall timeout

pablochacin replied:
@figo This is the time for the event to come after the clusterapi operation is executed (e.g. create a cluster). Not for the whole test to run.
That said,  would it work if I add an `async-timeout` flag to the test so different timeouts can be defined?  In order to do so, I must add a `TestMain` function and parse the parameter:

```
var asyncTimeout Duration

func TestMain(m *testing.M) { 
   duration := flag.DurationVar(&asyncTimeout, 10*time.Second, "async-timeout", "Asynchronous test timeout")
   flag.Parse() 

    retCode := m.Run()

    os.Exit(retCode)
}
```

figo replied:
@pablochacin i am suggesting to increase the timeout value, i understand it should not taking long from a operation to event been captured, but testing locally shows delay can happen (could be due to slow cluster).  the purpose of setting timeout probably is: event should be detected in a reasonable time, otherwise something went wrong in the cluster.  would 1minute been a reasonable time? or defer to overall test timeout (10 minutes?)

pablochacin replied:
Ok with one minute

pablochacin commented:
@ncdc @figo  I think I addressed all your concerns.

figo commented on test/integration/cluster/cluster_test.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+// Timeout for waiting events in seconds
+const TIMEOUT = 10
```
do you agree to 60? thanks

pablochacin replied:
Yes. Sorry! I should had changed it back to 10 for testing the async tests without waiting 60 seconds. Now is corrected.

figo commented:
/test pull-cluster-api-test

figo commented:
/lgtm

ncdc commented:
/lgtm
/approve

ncdc commented:
/assign @vincepri

vincepri commented on test/integration/cluster/cluster_test.go:
```diff
@@ -0,0 +1,129 @@
+}
+
+// Timeout for waiting events in seconds
+const TIMEOUT = 60
```
Should we make this a test flag with default to 60s?
